### ISSUE TITLE
Json.NET throws TypeLoadException on Blazor 0.7.0

### ISSUE THREAD

**Issue by danroth27** [mono/mono#11848]
_From @pgrho on November 17, 2018 1:37_

I'm updating my standalone Blazor app to use 0.7.0. And found the `JsonConvert.DeserializeObject<T>(string)` method of Newtonsoft.Json 11.0.2 throws `TypeLoadException` in the first invocation.  

## Repro steps

1. Create a New ASP.NET Core/Blazor Project. 
2. Reference Newtonsoft.Json 11.0.2
3. Add linker.xml

```
<linker>
    <assembly fullname="mscorlib">
        <!-- Preserve all methods on WasmRuntime, because these are called by JS-side code
    to implement timers. Fixes https://github.com/aspnet/Blazor/issues/239 -->
        <type fullname="System.Threading.WasmRuntime" />
    </assembly>
    <assembly fullname="System.Core">
        <!-- This is required by JSon.NET and any expression.Compile caller -->
        <type fullname="System.Linq.Expressions*" />
    </assembly>
    <!-- Name of the entry point assembly -->
    <assembly fullname="BlazorTest" />
</linker>
```

4. Reference linker.xml in the .csproj

```
  <ItemGroup>
    <BlazorLinkerDescriptor Include="Linker.xml" />
  </ItemGroup>
```

5. Invoke `JsonConvert.DeserializeObject<T>(string)` in some event.

```csharp
JsonConvert.DeserializeObject<EmptyClass>("{}");
```

The steps above will throw `TypeLoadException` for `INotifyPropertyChanged`, and this behavior won't change in the second call. But in my app Json.NET throws `TypeLoadException` for `INotifyPropertyChanging` and subsequent invocations works fine.

_Copied from original issue: aspnet/Blazor#1692_

**Comment by marek-safar** [comment]
This looks like we are loading wrong mscorlib. Could you attach linked output binaries?

**Comment by pgrho** [comment]
@marek-safar [BlazorTest.zip](https://github.com/mono/mono/files/2635464/BlazorTest.zip)

**Comment by DerekChasse** [comment]
Is there a workaround available? If we need this functionality should we downgrade Blazor to 0.6.0?

**Comment by pgrho** [comment]
@DerekChasse Use the interface explicitly before calling Json.NET.
```
public class Startup
{
    static Startup()
    {
        typeof(System.ComponentModel.INotifyPropertyChanging).GetHashCode();
        typeof(System.ComponentModel.INotifyPropertyChanged).GetHashCode();
    }
}
```

**Comment by pver** [comment]
Fyi, I had something similar, but with the System.Collections.Specialized.INotifyCollectionChanged interface.  Adding it to the workaround above solved the issue

**Comment by lewing** [comment]
I've tested this against mono/linker 16fe874 and it working fine.  There was quite a bit of work done on preserving interfaces since the build blazor is using in 0.7

**Comment by lewing** [comment]
This should be be fixed in the linker shipping with blazor 0.8.0

**Comment by FluentGuru** [comment]
I have the same issue but with blazor 0.9
but in mono cecil. If I disable the linker it build but then I get a similar error at runtime on the console.
WASM: Could not load signature of Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions:BuildServiceProvider due to: Could not resolve type with token 01000031 from typeref (expected class 'System.IAsyncDisposable' in assembly 'System.Runtime, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a') assembly:System.Runtime, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a type:System.IAsyncDisposable member:(null)
